File: py-langgraph/src/rag/retriever.py
"""
RAG 检索器 — Cloudflare Service（通过 Memory Gateway）
"""

from src.rag.vector_store import VectorStore


class Retriever:
    """RAG 检索器"""

    # 初始化当前对象
    def __init__(
        self,
        qdrant_url: str = "",
        collection_name: str = "documents",
        top_k: int = 5,
    ):
        # 保持接口兼容，qdrant_url 参数保留但不使用
        self._vector_store = VectorStore(
            url=qdrant_url,
            collection_name=collection_name,
        )
        self._top_k = top_k

    # 获取 retrieve 对应的数据
    async def retrieve(self, query: str, user_id: str) -> list[dict]:
        """检索相关文档"""
        return await self._vector_store.search(query, user_id=user_id, top_k=self._top_k)

    # 获取 retrieve with context 对应的数据
    async def retrieve_with_context(
        self,
        query: str,
        user_id: str,
        top_k: int | None = None,
    ) -> list[dict]:
        """检索并返回格式化的结果"""
        results = await self.retrieve(query, user_id)

        return [
            {
                "content": r["content"],
                "score": r["score"],
                "metadata": r.get("metadata", {}),
            }
            for r in results
        ]

File: ts-langchain/src/rag/retriever.ts
/**
 * RAG 检索器 — Cloudflare Service（通过 Memory Gateway）
 *
 * 封装文档向量搜索，对外提供统一接口。
 * 所有向量操作通过 Cloudflare Service Gateway 进行。
 */

import { VectorStore, SearchResult } from "./vector-store.js";

export interface RetrieverOptions {
  /** Cloudflare Memory Gateway 地址（可选） */
  baseUrl?: string;
  /** Gateway 认证密钥 */
  secret?: string;
  /** 返回条数 */
  topK?: number;
}

export class Retriever {
  private vectorStore: VectorStore;
  private topK: number;
  private userId: string;

  // 初始化检索器
  constructor(options: RetrieverOptions, userId: string) {
    this.vectorStore = new VectorStore({
      baseUrl: options.baseUrl,
      secret: options.secret,
    });
    this.topK = options.topK || 5;
    this.userId = userId;
  }

  /**
   * 检索相关文档
   */
  async retrieve(
    query: string,
    topK: number = this.topK,
  ): Promise<SearchResult[]> {
    return this.vectorStore.search(this.userId, query, topK);
  }

  /**
   * 批量检索
   */
  async retrieveBatch(
    queries: string[],
  ): Promise<Array<SearchResult & { query: string }>> {
    const results = await Promise.all(queries.map((q) => this.retrieve(q)));
    return results.flat().map((r, i) => ({
      ...r,
      query: queries[Math.floor(i / this.topK)],
    }));
  }
}
